execute/tokendata/usdc/http.go

```diff
@@ -0,0 +1,198 @@
+	attestationStatusPending attestationStatus = "pending_confirmations"
+)
+
+type httpResponse struct {
+	Status      attestationStatus `json:"status"`
+	Attestation string            `json:"attestation"`
+	Error       string            `json:"error"`
+}
+
+func (r httpResponse) validate() error {
+	if r.Error != "" {
+		return fmt.Errorf("attestation API error: %s", r.Error)
+	}
+	if r.Status == "" {
+		return fmt.Errorf("invalid attestation response")
+	}
+	if r.Status != attestationStatusSuccess {
+		return ErrNotReady
+	}
+	return nil
+}
+
+func (r httpResponse) attestationToBytes() ([]byte, error) {
+	attestationBytes, err := hex.DecodeString(strings.TrimPrefix(r.Attestation, "0x"))
+	if err != nil {
+		return nil, err
+	}
+	return attestationBytes, nil
+}
+
+func (h *httpClient) Get(ctx context.Context, messageHash [32]byte) ([]byte, int, error) {
+	var empty []byte
+	// Terminate immediately when rate limited
+	if h.inCoolDownPeriod() {
+		return empty, http.StatusTooManyRequests, ErrRateLimit
+	}
+
+	if h.rate != nil {
+		// Wait blocks until it the attestation API can be called or the
+		// context is Done.
+		if waitErr := h.rate.Wait(ctx); waitErr != nil {
+			return empty, http.StatusTooManyRequests, ErrRateLimit
+		}
+	}
+
+	timeoutCtx, cancel := context.WithTimeoutCause(ctx, h.apiTimeout, ErrTimeout)
+	defer cancel()
+
+	url := fmt.Sprintf("%s/%s/%s/0x%x", h.apiURL, apiVersion, attestationPath, messageHash)
+	return h.callAPI(timeoutCtx, url)
+}
+
+func (h *httpClient) callAPI(ctx context.Context, url string) ([]byte, int, error) {
+	var response []byte
+	// Use a timeout to guard against attestation API hanging, causing observation timeout and
+	// failing to make any progress.
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return response, http.StatusBadRequest, err
+	}
+	req.Header.Add("accept", "application/json")
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return response, http.StatusRequestTimeout, ErrTimeout
+		}
+		// On error, res is nil in most cases, do not read res.StatusCode, return BadRequest
+		return response, http.StatusBadRequest, err
+	}
+	defer res.Body.Close()
+
+	// Explicitly signal if the API is being rate limited
+	if res.StatusCode == http.StatusTooManyRequests {
+		h.setCoolDownPeriod(res.Header)
+		return response, res.StatusCode, ErrRateLimit
+	}
+	if res.StatusCode == http.StatusNotFound {
+		return response, res.StatusCode, ErrNotReady
+	}
+	if res.StatusCode != http.StatusOK {
+		return response, res.StatusCode, ErrUnknownResponse
+	}
+
+	return h.parsePayload(res)
+}
+
+func (h *httpClient) parsePayload(res *http.Response) ([]byte, int, error) {
+	var empty []byte
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return empty, res.StatusCode, err
+	}
+
+	var response httpResponse
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		return empty, res.StatusCode, err
+	}
```

Comment:
I think this can be simplified to use the `res.Body` read closer directly:

```suggestion
	var response httpResponse
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
	  return nil, res.StatusCode, fmt.Errorf("failed to decode json: %w", err)
	}
```

I also don't think you need `empty` you can just return `nil`